### PR TITLE
counters

### PR DESCRIPTION
**Pull request by aniketfuryrocks**

### CODE REVIEW

File: lite-rpc/src/main.rs [hunk 1]
```diff
@@ -31,6 +31,11 @@ async fn get_identity_keypair(identity_from_cli: &String) -> Keypair {
     }
 }
 
+lazy_static::lazy_static! {
+    static ref RESTARTS: IntCounter =
+    register_int_counter!(opts!("literpc_rpc_restarts", "Nutber of times lite rpc restarted")).unwrap();
```
Comment (godmodegalactus) [inline lite-rpc/src/main.rs:36]: nit! nutber

**Review by godmodegalactus — CHANGES_REQUESTED** [review]

File: core/src/solana_utils.rs [hunk 2]
```diff
@@ -61,7 +61,7 @@ impl SolanaUtils {
     }
 
     pub async fn poll_slots(
-        rpc_client: Arc<RpcClient>,
```
Comment (godmodegalactus) [inline core/src/solana_utils.rs:64]: Why did you remove arc?

Comment (aniketfuryrocks) [reply core/src/solana_utils.rs:64]: For the lifetime and type of function. It doesn't require the ownership of rpc_client. Therefore the reference of the inner of an Atomic Reference Counter.

File: lite-rpc/src/main.rs [hunk 3]
```diff
@@ -79,13 +75,52 @@ pub async fn main() -> anyhow::Result<()> {
...
+        || if let Ok(enable_postgres_env_var) = env::var("PG_ENABLED") {
+            enable_postgres_env_var != "false"
+        } else {
+            false
+        };
+
+    args
+}
+
+#[tokio::main]
```
Comment (godmodegalactus) [inline lite-rpc/src/main.rs:97]: I prefer it as before specifying 16 worker threads.

File: services/src/tpu_utils/tpu_service.rs [hunk 4]
```diff
@@ -163,48 +163,42 @@ impl TpuService {
...
-                    );
-                    break;
-                }
-            } else {
-                nb_errror = 0;
+                log::info!("Got error while polling slot {}", err);
             }
-        }
+
+            bail!("Reached max amount of errors to fetch latest slot, exiting poll slot loop")
```
Comment (godmodegalactus) [inline services/src/tpu_utils/tpu_service.rs:196]: Should also log an error! here

File: services/src/transaction_replayer.rs [hunk 5]
```diff
@@ -45,56 +42,44 @@ impl TransactionReplayer {
...
+                        error!("error while scheduling replay ({})", e);
+                        continue;
+                    } else {
+                        MESSAGES_IN_REPLAY_QUEUE.inc();
                     }
                 }
             }
-            Ok(())
+
+            bail!("transaction replay channel broken");
```
Comment (godmodegalactus) [inline services/src/transaction_replayer.rs:82]: here to add an explicit error! log

**Review by godmodegalactus — CHANGES_REQUESTED** [review]

File: services/src/transaction_replayer.rs (hunk 5 above)
Comment (aniketfuryrocks) [reply services/src/transaction_replayer.rs:82]: same here. It will be logged down the line

File: services/src/tpu_utils/tpu_service.rs (hunk 4 above)
Comment (godmodegalactus) [reply services/src/tpu_utils/tpu_service.rs:196]: You are sure right?

Comment (godmodegalactus) [reply services/src/tpu_utils/tpu_service.rs:196]: I prefer it having it twice rather than none because something went wrong with tokio stuff. Lite-rpc had problems in past where it went down without showing any logs which service went down.

**Comment by godmodegalactus** [comment]
Where is the prometheus variable that is 1 when RPC is replying and 0 when it is not?
Similarly, for postgres?

**Comment by aniketfuryrocks** [comment]
> 

It will restart if any of the 2 fail or any service fails

File: lite-rpc/src/main.rs [hunk 6]
```diff
@@ -79,13 +75,53 @@ pub async fn main() -> anyhow::Result<()> {
...
+    let args = get_args();
+
+    let main: AnyhowJoinHandle = tokio::spawn(async move {
+        loop {
+            let Err(err) = start_lite_rpc(args.clone()).await else{
+                bail!("LiteBridge services returned without error");
+            };
+
+            // log and restart
+            log::error!("Services quit unexpectedly {err:?} restarting in {RESTART_DURATION:?}");
```
Comment (aniketfuryrocks) [inline lite-rpc/src/main.rs:110]: It will log error here

Comment (godmodegalactus) [reply lite-rpc/src/main.rs:110]: Have you tested if it is working?

File: lite-rpc/src/bridge.rs [hunk 7]
```diff
@@ -57,6 +57,18 @@ lazy_static::lazy_static! {
     register_int_counter!(opts!("literpc_rpc_airdrop", "RPC call to request airdrop")).unwrap();
     static ref RPC_SIGNATURE_SUBSCRIBE: IntCounter =
     register_int_counter!(opts!("literpc_rpc_signature_subscribe", "RPC call to subscribe to signature")).unwrap();
+    static ref WS_SERVER_FAIL: IntCounter =
+    register_int_counter!(opts!("literpc_rpc_ws_server_fail", "WebSocket server failed")).unwrap();
+    static ref METRICS_SERVICE_FAIL: IntCounter =
+    register_int_counter!(opts!("literpc_rpc_metrics_service_fail", "Metrics service failed")).unwrap();
+    static ref HTTP_SERVER_FAIL: IntCounter =
+    register_int_counter!(opts!("literpc_rpc_http_server_fail", "Http server failed")).unwrap();
+    static ref PROMETHEUS_SERVER_FAIL: IntCounter =
```
Comment (aniketfuryrocks) [inline lite-rpc/src/bridge.rs:66]: Prometheus fail

**Comment by godmodegalactus** [comment]
> > 
> 
> It will restart if any of the 2 fail or any service fails. Log error and increment the restart counter. We can set alert when the restart counter increases.

It does not work. Simple we want a gauge that will go to 0 when triton rpc is working and will go to 0 if is not working. And it should again go to one when it is working again. So that we can display on dashboard that rpc is down or postgres is down etc. If you have questions discuss with Pan. Have you discussed with Pan about this idea and has he give a green light?

File: lite-rpc/src/main.rs (hunk 6 above)
Comment (aniketfuryrocks) [reply lite-rpc/src/main.rs:107]: Yes. Earlier it didn't log cause it was bailing out from the main. Which for some reason is not captured by Prometheus.

Comment (godmodegalactus) [reply lite-rpc/src/main.rs:107]: Ok, lets see then on fly if it works.

**Review by godmodegalactus — APPROVED** [review]
